packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,172 @@
+          !ts.isClassExpression(tsNode)
+        ) {
+          declEndPos = tsNode.body?.getStart() ?? tsNode.end;
+        }
+
+        for (const typeParameter of tsNode.typeParameters) {
+          const { uses } = usage.get(typeParameter.name)!;
+          let numExplicitUses = 0;
+          for (const use of uses) {
+            const pos = use.location.getStart();
+            if (pos > tsNode.getStart() && pos < declEndPos) {
+              numExplicitUses++;
+              if (numExplicitUses >= 2) {
+                break;
+              }
+            }
+          }
+          const inferredUses = inferredCounts?.get(typeParameter.name) ?? 0;
+          const numUses = numExplicitUses + inferredUses;
+          if (numUses === 1) {
+            context.report({
+              data: {
+                name: typeParameter.name.text,
+              },
+              node: parserServices.tsNodeToESTreeNodeMap.get(typeParameter),
+              messageId: 'sole',
+            });
+          }
+        }
+      },
+    };
+  },
+});
+
+// Fill in a counter of the number of times each type parameter appears in
+// the given type.
+function collectTypeParameterUsage(
+  checker: ts.TypeChecker,
+  rootType: ts.Type,
+  counts: Map<ts.Identifier, number>,
+): void {
+  const increment = (id: ts.Identifier): void => {
+    counts.set(id, 1 + (counts.get(id) ?? 0));
+  };
+
+  const process = (type: ts.Type): void => {
+    if (tsutils.isTypeParameter(type)) {
+      for (const decl of type.getSymbol()?.getDeclarations() ?? []) {
+        increment((decl as ts.TypeParameterDeclaration).name);
+        break;
+      }
+    }
+    // This should be "else if" but tsutils.isParameterType narrows type to never.
+    if (tsutils.isUnionOrIntersectionType(type)) {
+      type.types.forEach(process);
+    } else if (tsutils.isIndexedAccessType(type)) {
+      // This covers index access types like "T[K]"
+      process(type.objectType);
+      process(type.indexType);
+    } else if (tsutils.isTypeReference(type)) {
+      // This covers generic types like Map<K, V>.
+      for (const t of type.typeArguments ?? []) {
+        process(t);
+      }
+    } else if (tsutils.isObjectType(type)) {
+      // This covers inferred object types. This should be _after_
+      // isTypeReference to avoid descending into all the properties of a
+      // generic interface/class, e.g. Map<K, V>.
+      for (const sym of type.getProperties()) {
+        process(checker.getTypeOfSymbol(sym));
+      }
+    }
+    // TODO: mapped types, type predicate, conditional types, function types
+  };
+
+  process(rootType);
+}
+
+/** Count uses of type parameters in inferred return types. */
+function countInferredTypeParameterUsage(
+  checker: ts.TypeChecker,
+  tsNode: ts.SignatureDeclaration | ts.ClassLikeDeclaration,
+): Map<ts.Identifier, number> | null {
+  let inferredCounts: Map<ts.Identifier, number> | null = null;
+  const fnNodes = [];
```

Comment:
[Style] Nit: I find abbreviations hard to keep track of in my head, and have heard from non-native English speakers that it's harder for them in particular. Could you use full names when possible please?

```suggestion
  const functionNodes = [];
```

There's some leeway - `esNode` and `tsNode` would be annoying to be explicit on, and `for (i` is a good standard. But for unusual abbreviations here and elsewhere (`decl`, `fn`, `num`, ...) I think the request holds.